scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,895 @@
+    try:
+        model_info = next(
+            filter(lambda m: m["short-name"] == model_name, jsn["models"])
+        )
+    except StopIteration as e:
+        raise InvalidTargetModelName(name) from e
+
+    model_uuid = model_info["model-uuid"]
+    model_type = model_info["type"]
+
+    return Model(name=model_name, uuid=model_uuid, type=model_type)
+
+
+def try_guess_charm_type_name() -> Optional[str]:
+    """If we are running this from a charm project root, get the charm type name charm.py is using."""
+    try:
+        charm_path = Path(os.getcwd()) / "src" / "charm.py"
+        if charm_path.exists():
+            source = charm_path.read_text()
+            charms = re.compile(r"class (\D+)\(CharmBase\):").findall(source)
+            if len(charms) < 1:
+                raise RuntimeError(f"Not enough charms at {charm_path}.")
+            elif len(charms) > 1:
+                raise RuntimeError(f"Too many charms at {charm_path}.")
+            return charms[0]
+    except Exception as e:
+        logger.warning(f"unable to guess charm type: {e}")
+    return None
+
+
+class FormatOption(
+    str, Enum
+):  # Enum for typer support, str for native comparison and ==.
+    """Output formatting options for snapshot."""
+
+    state = "state"  # the default: will print the python repr of the State dataclass.
+    json = "json"
+    pytest = "pytest"
+
+
+def get_juju_version(juju_status: Dict) -> str:
+    """Get juju agent version from juju status output."""
+    return juju_status["model"]["version"]
+
+
+class RemoteUnitStateDB(UnitStateDB):
+    """Represents a remote unit's state db."""
+
+    def __init__(self, model: Optional[str], target: JujuUnitName):
+        self._tempfile = tempfile.NamedTemporaryFile()
+        super().__init__(self._tempfile.name)
+
+        self._model = model
+        self._target = target
+
+    def _fetch_state(self):
+        fetch_file(
+            self._target,
+            remote_path="./unit-state.db",
+            container_name="charm",
+            local_path=self._state_file,
+            model=self._model,
+        )
+
+    def _open_db(self) -> Optional[SQLiteStorage]:
+        if not self._has_state:
+            self._fetch_state()
+        return super()._open_db()
+
+
+def _snapshot(
+    target: str,
+    model: Optional[str] = None,
+    pprint: bool = True,
+    include: str = None,
+    include_juju_relation_data=False,
+    include_dead_relation_networks=False,
+    format: FormatOption = "state",
+    fetch_files: Dict[str, List[Path]] = None,
+    temp_dir_base_path: Path = SNAPSHOT_TEMPDIR_ROOT,
+):
+    """see snapshot's docstring"""
+
+    try:
+        target = JujuUnitName(target)
+    except InvalidTargetUnitName:
+        logger.critical(
+            f"invalid target: {target!r} is not a valid unit name. Should be formatted like so:"
+            f"`foo/1`, or `database/0`, or `myapp-foo-bar/42`."
+        )
+        exit(1)
```

Comment:
The `exit()` built-in is not actually "built-in" but provided by the [`site`](https://docs.python.org/3/library/site.html?highlight=site#module-site) module in the Python standard library. `site` is imported during initialization, but can be suppressed using the `-S` flag. This means I could break snapshot.py by running it with `python3 -S snapshot.py`. Generally it is only recommend to use `exit()` in the Python interpreter.

In your case here, it would be better to use `sys.exit(1)` from the sys module. That way you do not have to worry about people potentially suppressing the import of the `site` module. This applies throughout for all your exception handlers.

Reply:
oh, cool. Didn't know that.